### PR TITLE
spi: add support for variable chip-select polarity

### LINKED ISSUE #4133: Support for VirtualSpiMasterDevice with Active-High Chip Select
Currently we assume that virtual SPI devices expect an active-low chip-select pin. While this is the de-facto standard, it is not universally the case (e.g. the LPM0130m126 TFT display expects an active-high chip-select).

The way our types are currently structured, there is no way to generally model this behavior. Specifically, the `SpiMaster` trait has an associated type, `ChipSelect`, which the chip-select gpio embedded inside a `VirtualSpiDevice` must be of that type and the `VirtualSpiDevice` and its Mux toggles this chip high/low when selecting the relevant device to talk to. Unfortunately, unless the hardware implementation uses a `ChipSelect` that _happens_ to be generic (e.g. the nordics and sam4l do this, but none others), there is no way to correctly "wrap" the chip select in a type that reverses the polarity.

### PR DESCRIPTION
**Pull request by alevy**
### Pull Request Overview

This pull request fixes #4133.

SPI chip select can be either active-low (most common) or active-high (less common, but exists). This PR adds type infrastructure for components to require a certain polarity and for chip implementations of SPI to both assert via types which polarities they are compatible with (some chips can only do active-low while anything that just uses GPIOs for chip-select can do both) as well as to dynamically act based on the specified polarity.

### Testing Strategy

Mostly compile tested, along with end-to-end testing of a few specific SPI peripherals on specific nRF boards. There is very little chance for something to go horribly wrong here, except for getting the polarity wrong on some peripheral---and it should be active low (was the assumed default) everywhere except for the LPM013[...].

This would be a wonderful case for having automated testing of things like end-to-end SPI peripheral functionality on actual hardware!

### TODO or Help Wanted

- [X] Testing
- [X] Document new types

### Documentation Updated

- [X] Updated the relevant files in `/docs`, or no updates are required.

### Formatting

- [X] Ran `make prepush`.

### CODE REVIEW

File: kernel/src/hil/spi.rs
Comment (bradjc) [inline kernel/src/hil/spi.rs:1]: Would it be possible to change the trait for `SpiMaster::ChipSelect`? It would be nice to mirror the LED interface where the board simply provides a pin that does the correct thing for `activate()` and `deactivate()`.

Comment (alevy) [reply kernel/src/hil/spi.rs:1]: I agree with the preference, but I don't think it could apply here.

In particular, `ChipSelect` is an associated type of the `SpiMaster` trait, meaning each implementation of `SpiMaster` chooses exactly one concrete type. This is important because while some SPI controllers defer chip selection to the user, others do chip selection themselves and can only work with certain definition of `ChipSelect`. For example, the SAM4L SPI (not USART spi) and OpenTitan SPI have a fixed set of pins associated with SPI chip-select, and you tell the the SPI controller which one to use by writing an integer to a register.

Moreover, the same SPI controller may have different peripherals it talks to, one with active low and one with active high chip selects, so this can't be a universal decision for a spi controller.

So, as a result, is I don't think we can't use the type of `SPIMaster::ChipSelect` to distinguish between the chip select's polarity.

Comment (alevy) [reply kernel/src/hil/spi.rs:1]: @bradjc @brghena thoughts? ^

File: kernel/src/hil/spi.rs [hunk 1]
```diff
@@ -23,6 +23,52 @@ pub enum DataOrder {
...
+impl<P: crate::hil::gpio::Output> ChipSelectPin<P> {
+    pub fn new(pin: P, polarity: ChipSelectActivePolarity) -> Self {
+        Self { polarity, pin }
+    }
+}
+
+impl<P: crate::hil::gpio::Output + ?Sized> ChipSelectPin<P> {
+    pub fn activate(&self) {
+        self.polarity.deactivate(&self.pin);
+    }
```
Comment (brghena) [inline kernel/src/hil/spi.rs:65]: I think you meant to `activate()` here.

File: kernel/src/hil/spi.rs
Comment (brghena) [reply kernel/src/hil/spi.rs:1]: The SAM4L example makes sense to me. Otherwise, we could pass in pre-constructed `ChipSelectPin` types into SPI drivers from board files. But the `specify_chip_select()` function is actually necessary for some SPI peripherals.

**Review by brghena — COMMENTED** [review]
Looks good to me. One bug in the SPI HIL, but otherwise looks reasonable.

File: kernel/src/hil/spi.rs (hunk 1 above)
Comment (alevy) [reply kernel/src/hil/spi.rs]: Thanks! Fixed.

File: kernel/src/hil/spi.rs [hunk 2]
```diff
@@ -23,6 +23,52 @@ pub enum DataOrder {
     LSBFirst,
 }
 
+/// Clock polarity (CPOL) defines whether the SPI clock is high
+/// or low when idle.
```
Comment (bradjc) [inline kernel/src/hil/spi.rs:27]: ?

File: kernel/src/hil/spi.rs
Comment (bradjc) [reply kernel/src/hil/spi.rs:1]: I'm pretty sure I don't understand. Whatever type the SPI implementation uses to represent a chip select _must_ have some information about whether that chip select needs to be active high or active low. Rather than include that separately, we just require that it be part of the `Spi::ChipSelect` type.

So if the chip select is represented by a number and polarity isn't possible, something like:

```rust
struct MyIndexActiveLowChipSelect {
  index: u8
}

trait ChipSelectPolarity {
  fn polarity() -> Polarity
}

impl ChipSelectPolarity for MyIndexActiveLowChipSelect {
  fn polarity() -> Polarity {
    Polarity::Low
  }
}

let MySensorChipSelect = MyIndexActiveLowChipSelect {index: 1};
```

and

```rust
pub trait SpiMaster<'a> {
    type ChipSelect: ChipSelectPolarity + Copy;
    ...
}
```

If we can do polarity:

```rust
struct MyIndexChipSelect {
  index: u8,
  polarity: Polarity,
}

trait ChipSelectPolarity {
  fn polarity() -> Polarity
}

impl ChipSelectPolarity for MyIndexChipSelect {
  fn polarity() -> Polarity {
    self.polarity
  }
}

let MySensorChipSelect = MyIndexChipSelect {index: 1, polarity: Low};
```


That way `Spi::specify_chip_select()` still works as is.

The nice thing is the ChipSelect type defines what is possible. So instead of the SAM4L silently ignoring polarity, it isn't possible to construct a chip select with any polarity other than low (because it would set the CS to be type `MyIndexActiveLowChipSelect`).

**Comment by alevy** [comment]
@bradjc OK, let me try.

**Comment by alevy** [comment]
@bradjc Unforunately you're suggestion doesn't quite "just work."

Basically I agree with, what I think, the premise of your suggestion is:

We should make the type of ChipSelect match what's possible on the hardware rather than dynamically ignore/panic when a polarity request arrives that's not possible.

There are other requirements, I think which, when also incorporated don't quite allow this suggestion, and I'm not sure what the right design would be instead.

- Specific peripheral capsules should specify which polarity the chip select requires. i.e.  the LPM013[...] capsules knows that peripheral _requires_ chip select to be active high, while the rf233 capsules knows the rf233 requires chip select to be active low.

- Corollary, the board shouldn't be responsible for setting the chip select active polarity. *This is a possible point of difference that would simplify things

- All peripherals actually have such a requirement. In most cases this requirement is going to be active low (though this isn't specified anywhere as far as I can tell, it's just the state of practice).

- The virtualization layer(s) can't possibly set this polarity since they are responsible for coalescing multiple peripherals, each with different requirements.

- Ideally, an attempt to connect an active-high peripheral to a SPI controller (via virtualization) that only supports active-low would fail at compile time.

At the moment, the peripheral driver has no visibility into the chip select pin at _all_, it's not even nameable  by type parameters, since it's hidden inside the SPI master trait implementation's chip select associated type, which even accessible from the `SpiMasterDevice` trait.

Ideally, I think, there would either be a type constraint that a peripheral driver imposes that somehow bubbles down to the Spi controller telling it whether to set or clear when activating/deactivating chip select _or_ the peripheral driver communicates this dynamically. To do so without adding a parameter to one of the methods in `SpiMaster` would require something somewhere in the stack to have access to the chip select in a way that it can basically modify it's set polarity dynamically.

**Comment by bradjc** [comment]
tldr If we can get everything we want but effectively components have to set the CS polarity (rather than the capsule IC driver) that seems pretty good.

---

What hardware designers don't sometimes put not gates on their CS lines? :)

My mental model for CS is it _is_ board-specified, because it depends on which pin the chip is wired to, and isn't something the capsule driver knows. Now the polarity is realistically IC-specific. But coupling those, and having boards specify them, seems like a reasonable compromise.

> though this isn't specified anywhere as far as I can tell

We really have _two_ configurations which are implicit today: 1) whether the hardware supports different polarity and 2) what polarity an IC requires. Ideally we would encode both of those in the type system if we can.

> Ideally, an attempt to connect an active-high peripheral to a SPI controller (via virtualization) that only supports active-low would fail at compile time.

I was thinking this would fail effectively at code-writing-time, because the programmer wouldn't even be able to construct a valid CS type that specifies the CS line as active high. But yeah, the compiler would enforce that.

**Comment by alevy** [comment]
Pushed a new design that entirely avoids changing the HIL and instead just adds an opt-in wrapper type in the style suggested by @bradjc.

This is nice in the sense that it preserves the ability of a SPI controller that only supports active low or active high to avoid accepting any arguments that it might have to reject at runtime, however it still doesn't have any way to prevent hooking up an active-high peripheral to an active-low-only controller.

I also can't figure out a way to cleanly and generally have the component or peripheral capsule be responsible for choosing the polarity. I think I can do it _if_ the chip-select is required to be a `kernel::hil::spi::util::ChipSelect` struct...

File: boards/sma_q3/src/main.rs [hunk 3]
```diff
@@ -454,9 +454,9 @@ pub unsafe fn start() -> (
         );
 
         let disp_pin = &nrf52840_peripherals.gpio_port[Pin::P0_07];
-        let cs_pin = static_init!(
-            components::lpm013m126::Inverted<'static, nrf52840::gpio::GPIOPin>,
-            components::lpm013m126::Inverted(&nrf52840_peripherals.gpio_port[Pin::P0_05])
+        let cs_pin = kernel::hil::spi::util::ChipSelect::new(
+            &nrf52840_peripherals.gpio_port[Pin::P0_05],
+            kernel::hil::spi::util::ChipSelectActivePolarity::ActiveHigh,
```
Comment (brghena) [inline boards/sma_q3/src/main.rs:459]: You should have a comment here that loudly calls out that this SPI CS pin is ActiveHigh (lest someone copy-paste and have a big error)

File: kernel/src/hil/spi.rs [hunk 4]
```diff
@@ -40,6 +40,42 @@ pub enum ClockPhase {
...
+    pub enum ChipSelectActivePolarity {
+        ActiveLow,
+        ActiveHigh,
+    }
+
+    #[derive(Copy, Clone)]
+    pub struct ChipSelect<P> {
+        pub pin: P,
+        pub polarity: ChipSelectActivePolarity,
+    }
```
Comment (brghena) [inline kernel/src/hil/spi.rs:54]: Do you have an example of what this looks like if `pin` is not a GPIO pin, but rather some kind of opaque number?

**Comment by alevy** [comment]
Hear ye hear ye!

Pushed is a completely revised proposal that does not change the SPI trait at all and _does_ enforce compatibility between the SPI implementation and the peripheral's requirement (though, specified by the component, not the peripheral capsule). E.g. if a peripheral requires an active low chip select, it can be used with spi implementations that _only_ support active low, or with implementations that support both active high and low, etc.

The basic idea is to have a trait `IntoChipSelect` which is parameterized by the polarity requirement of the peripheral, and converts an underlying chip select type (a GPIO, a `u8`, whatever...) to a wrapper type used by the SPI implementation. A SPI implementation can then implement the trait for either active high, active low, or both.

The PR also adds a reusable wrapper type most useful for SPI implementations that simply rely on GPIO for chip select and can do both.

**Comment by alevy** [comment]
Remaining TODOs before converting to a regular PR (modulo any more feedback between now and then):

- Document all these new types
- Reconsider the type names and namespace, which may not be ideally ergonomic.

**Review by bradjc — COMMENTED** [review]
This is very cool!

File: kernel/src/hil/spi.rs [hunk 5]
```diff
@@ -40,6 +40,72 @@ pub enum ClockPhase {
...
+    impl ChipSelect for ActiveHigh {
+        const POLARITY: Polarity = Polarity::High;
+    }
+
+    pub trait IntoChipSelect<T, ChipSelectActivePolarity> {
+        fn into_cs(self) -> T;
+    }
+
+    #[derive(Copy, Clone)]
+    pub struct ChipSelectPolar<P> {
```
[the remark this answers is not in the document]
Comment (alevy) [reply kernel/src/hil/spi.rs:78]: This was intentional actually for the situation where a chip like the SAM4L that _does_ have some chip-specific way of naming the chip select pin (e.g. an index) but, unlike the SAM4L, does have a way to choose between active high and low (idk... may no such chip exists).

File: kernel/src/hil/spi.rs [hunk 6]
```diff
@@ -40,6 +40,72 @@ pub enum ClockPhase {
...
+        }
+
+        pub fn activate(&self) {
+            match self.polarity {
+                Polarity::Low => self.pin.clear(),
+                Polarity::High => self.pin.set(),
+            }
+        }
+    }
+}
```
[the remark this answers is not in the document]
Comment (alevy) [reply kernel/src/hil/spi.rs:165]: This would have to be with the assumption that those controllers _all_ implement active low _only_ (likely?). See the revised version of SAM4L, which I think is better anyway and requires the implementation to declare explicitly what polarity it supports.

File: kernel/src/hil/spi.rs (hunk 4 above)
Comment (alevy) [reply kernel/src/hil/spi.rs]: It would rely on there being some HW register or something that sets the polarity, but:

```rust
impl SpiMaster<...> for MySpi {
    type ChipSelect = ChipSelect<u8>;

    ...

    fn specify_chip_select(&self, cs: Self::ChipSelect) {
        self.registers.csid.set(cs.pin);
        self.registers.cspolarity.set(match self.polarity {
              Polarity::ActiveLow => regs::CSPOL::LOW,
              Polarity::ActiveHigh => regs::CSPOL::HIGH,
        });
    }
}
```

File: kernel/src/hil/spi.rs [hunk 7]
```diff
@@ -40,6 +40,133 @@ pub enum ClockPhase {
...
+    }
+
+    impl ChipSelectActivePolarity for ActiveHigh {
+        const POLARITY: Polarity = Polarity::High;
+    }
+
+    /// A type that can be converted to the appropriate type for
+    /// [SpiMaster::ChipSelect](super::SpiMaster::ChipSelect) for a
+    /// particular `POLARITY`.
+    ///
```
Comment (bradjc) [inline kernel/src/hil/spi.rs:89]:
```suggestion
    /// A type that can be converted to the appropriate type for
    /// [SpiMaster::ChipSelect](super::SpiMaster::ChipSelect) for a
    /// particular `POLARITY`.
    ///
    /// Instantiating a driver for any SPI peripheral should require a type that
    /// implements [IntoChipSelect]. That enforces that whatever object is used
    /// as the chip select can support the correct polarity for the particular SPI
    /// peripheral. This is mostly commonly handled by the component for the
    /// peripheral, which requires an object with type [IntoChipSelect] and then
    /// converts the object to the [SpiMaster::ChipSelect](super::SpiMaster::ChipSelect)
    /// type.
```

File: kernel/src/hil/spi.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (bradjc) [reply kernel/src/hil/spi.rs:126]: I think this is already pretty advanced, and more handholding at the expense of a little code redundancy is worth it. Having this as only for both polarity GPIO-based chip selects makes it a little simpler to know how to implement this SPI interface for a new chip (ie whether or not to use `ChipSelectPolar`). But, I don't feel strongly about this and having a general type is fine.

**Review by bradjc — COMMENTED** [review]
Looks good!

File: kernel/src/hil/spi.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (alevy) [reply kernel/src/hil/spi.rs:126]: OK, I'm convinced. It's also easy to relax that constraint if ever we find an actual SPI controller that would want this (currently there are none, though waiting on confirmation for OpenTitan).

File: kernel/src/hil/spi.rs [hunk 8]
```diff
@@ -40,6 +40,139 @@ pub enum ClockPhase {
...
+    /// high chip select pins, in which case, they should implement
+    /// both the [ActiveLow] and [ActiveHigh] variants, for example,
+    /// using the [ChipSelectPolar] wrapper struct (which implements
+    /// both).
+    pub trait IntoChipSelect<T, POLARITY> {
+        fn into_cs(self) -> T;
+    }
+
+    /// A convenience wrapper type around
+    /// [Output](crate::hil::gpio::Output] GPIO pins that implements
```
Comment (bradjc) [inline kernel/src/hil/spi.rs:127]:
```suggestion
    /// [Output](crate::hil::gpio::Output) GPIO pins that implements
```

**Review by bradjc — DISMISSED** [review]

**Review by lschuermann — COMMENTED** [review]
This looks reasonable to me. It'd be awesome if we could remove the reference-trait implementation, or add a comment for why it's required.

**Review by ppannuto — DISMISSED** [review]
This is really cool! A great example of type system helping out with getting hardware configuration correct

File: kernel/src/hil/gpio.rs [hunk 9]
```diff
@@ -176,6 +176,20 @@ pub trait Output {
     }
 }
 
+impl<'a, O: Output> Output for &'a O {
```
[the remark this answers is not in the document]
Comment (alevy) [reply kernel/src/hil/gpio.rs:179]: Well, specifically, `ChipSelectPolar` needs to be able to wrap a type `&'a SOME_SPECIFIC_GPIO` _and_ `ChipSelectPolar` needs to be `Copy`---a constraint imposed by the `type ChipSelect` associated type in `SpiMaster` that's used in the virtual spi stack, though maybe can be worked around.

We could have `ChipSelectPolar<'a, P: gpio::Output>`, but then it won't be `Copy` in cases where the actual type of `P` is not Copy, which is most GPIO implementations (because they typically contain a client which is not Copy). Even if the field is a `&'a P`, the struct still can't be marked Copy, and it won't be usable as a `ChipSelect`.

In other words, the `ChipSelectPolar` needs to name only types that are Copy, which excludes most implementations of `gpio::Output`.

I'm open to suggestions!

[the remark this answers is not in the document]
Comment (alevy) [reply kernel/src/hil/gpio.rs]: OK, removed by making `ChipSelectPolar` take a lifetime parameter and always store a reference.

[the remark this answers is not in the document]
Comment (lschuermann) [reply kernel/src/hil/gpio.rs]: @alevy and I discussed this offline, and for posterity's sake, this is an issue because `ChipSelectPolar` wants to take either an owned struct that is `Copy`, or a shared reference (which is `Copy` by definition, despite the compiler's `Copy` derive not determining the trait bounds correctly: https://github.com/rust-lang/rust/issues/26925).

What we'd want from the compiler here is one of the following:
- either allow accepting a reference `&T` where `T: Trait` as a type parameter `P: Trait` (although that would probably open a whole different can of worms),
- or, more conservatively, the ability to accept some `P: Borrow<impl Trait>`. Because owned types generally allow to be borrowed as references to themselves, this trait bound would admit both owned types and references to them.

  Notably, this is already supported, but its ergonomics are quite bad. You can't use `impl Trait` as a type parameter for generic traits and have the compiler plug in the actual type automatically. Instead, you have to write a signature like `T: Trait, P: Borrow<T>`. However, the spurious type parameter `T` will then need to be explicitly specified each time, and also be "used" by introducing a `PhantomData<T>`.

It seems "wrong" to propagate these issues towards the `Output` trait, which `ChipSelectPolar` merely consumes. Notably, if these structs and traits were defined in different crates, an `impl<T: Output> Output for &T { ... }` would not be admitted in the `ChipSelectPolar`-defining crate -- it would require changes to the upstream crate.

Simply always taking a reference works around these issues.

**Review by lschuermann — APPROVED** [review]

**Review by bradjc — APPROVED** [review]
